expression/builtin_cast_vec.go

```diff
@@ -143,3 +143,49 @@ func (b *builtinCastRealAsRealSig) vecEvalReal(input *chunk.Chunk, result *chunk
 func (b *builtinCastRealAsRealSig) vectorized() bool {
 	return true
 }
+
+func (b *builtinCastDecimalAsRealSig) vecEvalReal(input *chunk.Chunk, result *chunk.Column) error {
+	n := input.NumRows()
+	buf, err := b.bufAllocator.get(types.ETDecimal, n)
+	if err != nil {
+		return err
+	}
+	defer b.bufAllocator.put(buf)
+	if err := b.args[0].VecEvalDecimal(b.ctx, input, buf); err != nil {
+		return err
+	}
+
+	result.ResizeFloat64(n, false)
+	result.MergeNulls(buf)
+
+	d := buf.Decimals()
+	rs := result.Float64s()
+
+	inUnionAndUnsigned := b.inUnion && mysql.HasUnsignedFlag(b.tp.Flag)
+	for i := 0; i < n; i++ {
+		if result.IsNull(i) {
+			continue
+		}
+		if inUnionAndUnsigned && d[i].IsNegative() {
+			rs[i] = 0
+		} else {
+			res, err := d[i].ToFloat64()
+			if err != nil {
+				if types.ErrOverflow.Equal(err) {
+					err = b.ctx.GetSessionVars().StmtCtx.HandleOverflow(err, err)
+				}
+				if err != nil {
```

Comment:
we should do this check only when `err` is changed in `err = b.ctx.GetSessionVars().StmtCtx.HandleOverflow(err, err)`

Reply:
I'm a little confused. HandleOverflow treats ErrOverflow as warnings (return nil) or returns the error (the err). if HandleOverFlow return the err, how should we handle it?

Reply:
`HandleOverflow()` tries to convert the error to warnings according to certain sql mode. It may return that error in some scenarios. You can see the internal implementation of that function for more detail.